Allow custom UserDefaults object, bug fixed

albertwujj opened the pull request:
Added an optional parameter to Zephyr.sync(), taking a custom UserDefaults object and using that in place of always using the container-wide UserDefaults.standard.

Fixed bug resulting in UserDefaults.standard being used despite passing in a custom UserDefaults object to the variable-arguments sync() function (when no keys were passed in)

albertwujj committed:
Allowed custom UserDefaults object

albertwujj committed:
Passed optional parameter to overloaded function

sync(keys: keys) =>
sync(userDefaults: userDefaults, keys: keys)

ArtSabintsev commented:
Please cleanup all the whitespace issues. Makes it really hard to see the diff. It's probably space/tabs issue in your Xcode settings.

I will review it afterwards.

albertwujj committed:
Trimmed whitespaces in blank lines

albertwujj commented:
Hey, I trimmed whitespaces from blank lines with Xcode settings. Thanks for the tip!

albertwujj commented:
Hey, I trimmed the whitespaces from the blank lines. Thanks for the tip!

albertwujj committed:
Update Zephyr.swift

albertwujj committed:
Matched whitespaces

albertwujj commented:
By the way, some lines with "///" or "/// - Parameters:" had an extra space in front of them, some didn't. I matched them anyways. Kinda satisfying.

ArtSabintsev commented:
Alright, I need to think about a few things with respect to this PR, specifically with the fact that this is introducing a breaking API change and will involve some `@deprecated` flags.

Was there an issue creating a new `sync` method that added the userDefaults param, and for the default scenario, caused down to the original `sync` method?

albertwujj commented:
> for the default scenario, caused down to the original sync method?

Not sure what exactly you mean. But I didn't create any new methods. I only added an optional parameter. However, you had two `sync` methods, one that called the other, and I forgot to pass
the optional parameter from one method to the other.

ArtSabintsev commented:
I think you should create a third sync method that calls down to the others after it does what it needs to do. Doing that makes it so the API doesn't have a breaking change.

albertwujj committed:
Added sync func for custom UserDefaults

Added 3rd and 4th overloaded sync func, with a custom UserDefaults object as an optional parameter, rather than adding the optional parameter to the original functions.

albertwujj commented:
Got it. I committed changes adding the 3rd and 4th overloaded `sync` functions (tested).

On another note, I am curious, why did you have the variadic-argument `sync` function deal with the empty `keys` case rather than have the array-argument `sync` function deal with both empty and non-empty cases?

albertwujj commented:
To clarify though, the UserDefaults parameter was an optional parameter so, in my previous commit, one could still call the `sync` function as normal without passing in a UserDefaults object. The default value of the optional parameter was UserDefaults.standard.

ArtSabintsev commented:
I haven't forgotten about this PR. I will look later this week.

ArtSabintsev commented on Sources/Zephyr.swift:
```diff
@@ -46,14 +46,17 @@ public class Zephyr: NSObject {
         return NSUbiquitousKeyValueStore.default.dictionaryRepresentation
     }
 
+    //The UserDefaults object to sync with NSUbiquitousKeyValueStore/iCloud
```
Three `/` and spacing afterwards.

ArtSabintsev replied:
Actually, I'll merge it in and make the change myself and release this as part of the Swift 4.1 release I'm about to cut.

Thanks!

ArtSabintsev reviewed (approved):
Looks good. Just make that one comment change and I'll merge this in.

Thanks!

ArtSabintsev commented:
Merged and release in v3.1.0.

Thanks again!
